[Lin Bin] iP

linnn-7 opened the pull request:
### **Miss-A**

> “Soon is not as good as now.” – Seth Godin([source](https://dansilvestre.com/productivity-quotes))

Miss-A relieves your mind from the burden of remembering tasks you need to do. It's:
- text-based
- very friendly for beginners
- _VERY_ SMART

All you need to do is:
1. download it from [here](https://github.com/linnn-7/ip/releases/download/A-Jar/missa.jar).
2. open it using command java -jar missa.jar
3. add your tasks.

let it manage your tasks for you 🙋‍♀️
And it is **FREE**! 💪

Features:
- [X] Managing tasks (todo, deadline, event)
- [X]  Searching tasks
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    MissA missA = new MissA("src/main/java/data/tasks.txt");
    missA.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

linnn-7 committed:
initial version that greets and exit

linnn-7 committed:
echos commands entered by the user, and exits when the user types the command bye.

linnn-7 committed:
store text entered by the user and display them back

linnn-7 committed:
Add the ability to mark tasks as done and the ability to change the status back to not done.

linnn-7 committed:
Add support for tracking three types of tasks

linnn-7 committed:
Use Exceptions to handle errors

linnn-7 committed:
Add support for deleting tasks from the list.

linnn-7 committed:
Test using the I/O redirection technique

linnn-7 committed:
Update Javadoc according to Git conventions

linnn-7 committed:
Save the tasks in the hard disk automatically

linnn-7 committed:
Merge branch 'branch-level-7'

* branch-level-7:
  Save the tasks in the hard disk automatically

linnn-7 committed:
Store deadline and event date and time

linnn-7 committed:
Merge branch 'branch-level-8'

* branch-level-8:
  Store deadline and event date and time

linnn-7 committed:
Refactor the code to extract out closely related code as classes

linnn-7 committed:
Divide classes into packages

linnn-7 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

* origin/add-gradle-support:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

linnn-7 committed:
Automate project builds using Gradle

linnn-7 committed:
Add JUnit tests to test the behavior of the code

linnn-7 committed:
Add JavaDoc comments

linnn-7 committed:
Tweak the code to comply with a coding standard

linnn-7 committed:
Allow users to find a task by searching for a keyword

linnn-7 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

linnn-7 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with a coding standard

linnn-7 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Allow users to find a task by searching for a keyword

linnn-7 committed:
Fix Gradle checkstyle bugs

linnn-7 committed:
Use checkStyle to detect coding style violations

linnn-7 committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Use checkStyle to detect coding style violations

linnn-7 committed:
Add a GUI to the chatbot

linnn-7 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add a GUI to the chatbot

chonghaoen commented on src/main/java/missa/Main.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
```
Good job following the correct naming convention for abbreviations! 💯

chonghaoen commented on src/main/java/missa/MissA.java:
```diff
@@ -0,0 +1,86 @@
+            String newData = tasks.getUpdatedData();
+            storage.writeBackData(newData);
+            return command.getReply(ui);
+        } catch (IncorrectTaskTypeException
+                 | NoSuchTaskException
+                 | NoTimingException
+                 | NoContentException e) {
```
Good job splitting long lines of code for visual clarity! 💯

chonghaoen commented on src/main/java/missa/Parser.java:
```diff
@@ -0,0 +1,153 @@
+                String content = task[1];
+                nextTask = new ToDo(content);
+                return new AddCommand(nextTask, tasks);
+
+            } else if (taskType.equals("deadline")) { // Checks if the task type is deadline.
+                if (task.length < 2) {
+                    throw new NoContentException();
+                }
+                if (!task[1].contains("/by")) {
+                    throw new NoTimingException();
+                }
+                String[] content = task[1].split(" /by ");
+
+                nextTask = new Deadline(
+                        content[0],
+                        strToDateTime(content[1]));
+                return new AddCommand(nextTask, tasks);
+
+            } else if (taskType.equals("event")) { // Checks if the task type is event.
+                if (task.length < 2) {
+                    throw new NoContentException();
+                }
+                if (!task[1].contains("/from") || !task[1].contains("/to")) {
+                    throw new NoTimingException();
+                }
+                String[] content = task[1].split(" /from ");
+                String text = content[0];
+                String[] interval = content[1].split(" /to ");
+
+                nextTask = new Event(
+                        text,
+                        strToDateTime(interval[0]),
+                        strToDateTime(interval[1]));
+                return new AddCommand(nextTask, tasks);
+
+            } else {
+                throw new IncorrectTaskTypeException();
+            }
+        }
+    }
```
Would this look clearer and easier to read by using switch-case? 🤔

chonghaoen commented on src/main/java/missa/Ui.java:
```diff
@@ -0,0 +1,126 @@
+    private String emptyLine = "________________________________________\n";
+
+    /** Greeting sentences */
+    private String greeting = "Hello from Miss A\n"
+            + "What can I do for you?\n"
+            + emptyLine
+            + "I can record 3 types of tasks now.\n"
+            + "ToDos: e.g. todo clean my room\n"
+            + "Deadlines: e.g. deadline submit homework /by 2021-4-1 21\n"
+            + "Events: e.g. event lecture /from 2021-4-2 14 /to 2021-4-2 16\n"
+            + emptyLine
+            + "Here are the commands that I can understand:)\n"
+            + "\"list\": I will display the latest task list.\n"
+            + "\"mark/unmark + number\": I will mark/unmark task in the list!\n"
+            + "\"delete + number\": I will remove task from the list.\n"
+            + "\"find + keyword\": I will search for relevant tasks recorded!\n"
+            + "\"bye\": this program will be closed.\n";
```
I like how easy this is to read! Would consider using this style in my code too!

chonghaoen reviewed (commented):
LGTM. Follows the coding standard closely and accurately.

linnn-7 committed:
Change GUI appearance

Changing GUI appearance makes instructions and user guides more clear for users to see.

linnn-7 committed:
Use Assertions

Using assertions documents assumptions such as taskList must be initialised.

linnn-7 committed:
Improve code quality

Let's,
* shorten long methods to improve readabilty
* avoid deep nesting to make logic obvious

linnn-7 committed:
Merge pull request #1 from linnn-7/branch-A-Assertions

Use Assertions

linnn-7 committed:
Revert "Use Assertions"

linnn-7 committed:
Merge pull request #2 from linnn-7/revert-1-branch-A-Assertions

Use Assertions

linnn-7 committed:
Add more assertions

linnn-7 committed:
Merge pull request #4 from linnn-7/branch-A-Assertions

Add assertions

linnn-7 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add more assertions
  Revert "Use Assertions"
  Use Assertions

linnn-7 committed:
Merge pull request #3 from linnn-7/branch-A-CodeQuality

Improve code quality

linnn-7 committed:
Merge branch 'master' of https://github.com/linnn-7/ip

* 'master' of https://github.com/linnn-7/ip:
  Improve code quality

linnn-7 committed:
Support adding tasks that have a prerequisite

linnn-7 committed:
Merge branch 'branch-B-DoAfterTasks'

* branch-B-DoAfterTasks:
  Support adding tasks that have a prerequisite

linnn-7 committed:
Improve the GUI

linnn-7 committed:
Add a representative screenshot

linnn-7 committed:
Add a User Guide

linnn-7 committed:
Change image link

linnn-7 committed:
Improve code quality

linnn-7 committed:
Fix minor bugs

linnn-7 committed:
Fix release version bugs

linnn-7 committed:
Auto-create data file for user

linnn-7 committed:
Update UI
